extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/common/GcpCredentials.java

```diff
@@ -0,0 +1,138 @@
+     * Otherwise it will be retrieved from a token or a Google Credentials file
+     *
+     * @return GoogleCredentials
+     */
+    public GoogleCredentials resolveGoogleCredentialsFromDataAddress(GcpServiceAccountCredentials gcpServiceAccountCredentials) {
+        var vaultTokenKeyName = gcpServiceAccountCredentials.getVaultTokenKeyName();
+        var vaultServiceAccountKeyName = gcpServiceAccountCredentials.getVaultServiceAccountKeyName();
+        var serviceAccountValue = gcpServiceAccountCredentials.getServiceAccountValue();
+
+        if (!StringUtils.isNullOrBlank(vaultTokenKeyName)) {
+            var token = vault.resolveSecret(vaultTokenKeyName);
+            if (StringUtils.isNullOrEmpty(token)) {
+                throw new GcpException(vaultTokenKeyName + " could not be retrieved from the vault.");
+            }
+            return createGoogleCredential(token, GcpCredentialType.GOOGLE_ACCESS_TOKEN);
+        } else if (!StringUtils.isNullOrBlank(vaultServiceAccountKeyName)) {
+            var token = vault.resolveSecret(vaultServiceAccountKeyName);
+            if (StringUtils.isNullOrEmpty(token)) {
+                throw new GcpException(vaultServiceAccountKeyName + " could not be retrieved from the vault.");
+            }
+            return createGoogleCredential(token, GcpCredentialType.GOOGLE_SERVICE_ACCOUNT_KEY_FILE);
+        } else if (!StringUtils.isNullOrBlank(serviceAccountValue)) {
+            var serviceKeyContent = new String(b64Decoder.decode(serviceAccountValue));
+            if (!serviceKeyContent.contains(SERVICE_ACCOUNT)) {
+                throw new GcpException("SERVICE_ACCOUNT_VALUE is not provided as a valid service account key file.");
+            }
+            return createGoogleCredential(serviceKeyContent, GcpCredentialType.GOOGLE_SERVICE_ACCOUNT_KEY_FILE);
+        } else {
+            return createApplicationDefaultCredentials();
+        }
+    }
+
+    /**
+     * Returns the Google Credentials which will created based on the Application Default Credentials in the following approaches
+     * - Credentials file pointed to by the GOOGLE_APPLICATION_CREDENTIALS environment variable
+     * - Credentials provided by the Google Cloud SDK gcloud auth application-default login command
+     * - Google App Engine built-in credentials
+     * - Google Cloud Shell built-in credentials
+     * - Google Compute Engine built-in credentials
+     *
+     * @return GoogleCredentials
+     */
+    public GoogleCredentials createApplicationDefaultCredentials() {
+        return createGoogleCredential("", GcpCredentialType.DEFAULT_APPLICATION);
+    }
+
+
+    public GoogleCredentials createGoogleCredential(String keyContent, GcpCredentialType gcpCredentialType) {
+        Objects.requireNonNull(keyContent, "key content");
+        GoogleCredentials googleCredentials;
+
+        if (gcpCredentialType.equals(GcpCredentialType.GOOGLE_ACCESS_TOKEN)) {
```

Comment:
this method seems quite big, please consider extracting methods from the different `if` branches, and the `if` itself could be replaced by a `switch case` statement that in my opinion would be more readable (and you could use the `return switch` form that's quite nice.